ui/segmented-bar.reel/segmented-bar.js

```diff
@@ -0,0 +1,132 @@
+    },
+    /**
+     * contentController accepts a 'RangeController' or 'TreeController', the
+     * contentController will default to a RangeController. The TreeController
+     * serves to flatten out a tree to be portrayed as a bar.
+     */
+    contentController: {value: null},
+
+    labelKeyName: {value: null},
+    valueKeyName: {value: null},
+    remainingLabelName: {value: null},
+    remainingValue: {value: null},
+
+    _maxValue: {value: null},
+    maxValue: {
+        get: function () {return this._maxValue},
+        set: function (value) {
+            this._maxValue = value;
+        }
+    },
+
+    /**
+     * Orientation of the bar will be horizontal by default.
+     * The orientation is set to automatically adjust according to the value
+     * of the screen width.
+     */
+    orientation: {value: null},
+    /**
+     * Allows user to override orientation.
+     * TODO: needed?
+     */
+    orientationOverride: {value: null},
+    /**
+     * TODO: needed?
+     */
+    orientationBreakpoint: {value: null},
+
+    /**
+     * logic for RangeController
+     * accepts an RangeController to be projected into the SegmentedBar
+     * TODO: needed? or just an array?
+     */
+
+    /**
+     * logic to flatten TreeController
+     * serves to flatten out a tree to be projected into the SegmentedBar
+     */
+    _flattenTreeController: {value: function () {}},
+
+    enterDocument: {value: function () {
+        this._constructRemainingSegment();
+        // TODO: perf review of listening to win event here
+        // TODO: why doesn't this work in prepareForActivationEvents?
+        window.addEventListener("resize", this);
+        this.orientation =
+            window.matchMedia("(max-width: 1000px)").matches ? "vertical": "horizontal";
+    }},
+    exitDocument: {value: function () {
+        // TODO: what if another component needs to listen for this?
+        window.removeEventListener("resize", this);
+    }},
+
+    /**
+     * Calculate remaining value & construct remaining segment
+     * The remaining value is then pushed into this._content array
+     * to be projected as a segment of the SegmentedBar
+     */
+    _constructRemainingSegment: {
+        value: function _constructRemainingSegment() {
+            var sum = 0;
+            for (var i = 0; i < this.content.length; i++) {
+                sum += this.content[i][this.valueKeyName];
+            }
+
+            // safe guard against user passing in values that total to be greater than max
+            if (sum > this.maxValue) {
+                console.warn(
+                    "SegmentedBar content values total " + sum +
+                    ", maxValue overridden from " + this.maxValue +
+                    " to match.");
+                this.maxValue = sum;
+            } else {
+                this.remainingValue = this.maxValue - sum;
+
+                // construct remaining segment
+                var remainingSegment = {};
```

Comment:
I thought we had agreed that all values are expected to be in the array of data passed to the segmentbar so that we wouldn't have to deal with that, make the contract simpler and consitent, no?